File: go/quasiquote.go
package tabgo

func QuasiquoteList(ast Tab) Tab {
	result := ListToTab(TabList{})
	astList := ToList(ast)
	for i := len(astList) - 1; i >= 0; i-- {
		elt := astList[i]
		if IsList(elt) &&
			len(ToList(elt)) > 0 &&
			IsSymbol(ToList(elt)[0]) &&
			ToSymbol(ToList(elt)[0]) == "..unq" {
			result = ListToTab(TabList{
				SymbolToTab("concat"),
				ToList(elt)[1],
				result,
			})
		} else {
			result = ListToTab(TabList{
				SymbolToTab("cons"),
				Quasiquote(elt),
				result,
			})
		}
	}
	return result
}

func Quasiquote(ast Tab) Tab {
	if IsList(ast) &&
		len(ToList(ast)) > 0 &&
		IsSymbol(ToList(ast)[0]) &&
		ToSymbol(ToList(ast)[0]) == "unq" {
		return ToList(ast)[1]
	}
	if IsList(ast) {
		return QuasiquoteList(ast)
	}
	if IsDict(ast) || IsSymbol(ast) {
		return ListToTab(TabList{
			SymbolToTab("q"),
			ast,
		})
	}
	return ast
}

File: js/quasiquote.js
import { TalSymbol, TalMap } from './mal-types.js';

const quasiquoteList = (ast) => {
	let result = [];
	for(let i = ast.length - 1; i >= 0; i--) {
		const elt = ast[i];
		if(
			elt instanceof Array &&
			elt.length > 0 &&
			elt[0] instanceof TalSymbol &&
			elt[0].valueOf() === '..unq'
		) {
			result = [
				new TalSymbol('concat'),
				elt[1],
				result,
			];
		} else {
			result = [
				new TalSymbol('cons'),
				quasiquote(elt),
				result,
			];
		}
	}
	return result;
};

export const quasiquote = (ast) => {
	if(
		ast instanceof Array &&
		ast.length > 0 &&
		ast[0] instanceof TalSymbol &&
		ast[0].valueOf() === 'unq'
	) {
		return ast[1];
	}
	if(ast instanceof Array) {
		return quasiquoteList(ast);
	}
	if(ast instanceof TalMap || ast instanceof TalSymbol) {
		return [
			new TalSymbol('quote'),
			ast,
		];
	}
	return ast;
};